File: Data Structures & Algorithms/swim-in-rising-water/submission-5.java
class Solution {
    public int swimInWater(int[][] grid) {
        int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        int m = grid.length;
        int n = grid[0].length;
        int[][] visited = new int[m][n];
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        pq.add(new int[] {grid[0][0], 0, 0});
        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int time = cur[0];
            int r = cur[1];
            int c = cur[2];
            if (r == m - 1 && c == n - 1) {
                return time;
            }
            for (int[] dir : directions) {
                int newR = r + dir[0];
                int newC = c + dir[1];
                if (newR >= 0 & newR < m && newC >= 0 && newC < n) {
                    if (visited[newR][newC] == 0) {
                        pq.add(new int[] {Math.max(time, grid[newR][newC]), newR, newC});
                        visited[newR][newC] = 1;
                    }
                }
            }
        }
        return -1;

    }
}
 

File: Data Structures & Algorithms/count-number-of-islands/submission-5.cpp
class Solution {
public:
    vector<vector<int>> directions = {{1,0}, {0,1}, {-1,0}, {0,-1}};
    int numIslands(vector<vector<char>>& grid) {
        int count = 0;
        int m = grid.size();
        int n = grid[0].size();
        for (int i = 0; i < m; i++) {
            for (int k = 0; k < n; k++) {
                if (grid[i][k] == '1') {
                    dfs(grid, i, k);
                    count++;
                }
            }
        }
        return count;

    }

    void dfs(vector<vector<char>>& grid, int r, int c) {
        int m = grid.size();
        int n = grid[0].size();
        grid[r][c] = '0';
        for (const auto& dir: directions) {
            int newR = r + dir[0];
            int newC = c + dir[1];
            if (newR >= 0 && newR < m && newC >= 0 && newC < n && grid[newR][newC] == '1') {
                dfs(grid, newR, newC);
            }
        }

    }
};
